datafusion/src/physical_plan/hash_join.rs

```diff
@@ -333,8 +334,10 @@ impl ExecutionPlan for HashJoinExec {
                                 concat_batches(&self.left.schema(), &batches, num_rows)?;
 
                             let left_side = Arc::new((hashmap, single_batch));
-
-                            *build_side = Some(left_side.clone());
+                            {
```

Comment:
Looking at the code, it seems like by moving the acquisition of the lock down, here multiple partitions could potentially race to build the `build_side` hash map. 

So I don't fully understand how this change is increasing performance. I need to think about this more 🤔 

cc @Dandandan

Reply:
Same here, I think this will result in all partitions to execute the same collect left & hashing compute. I am surprised this improves the performance, if anything it should make things slower?

Reply:
The same, also surprised. Didn't yet benchmark this, are you sure the numbers are correct @xudong963 ?

Reply:
I doubt myself a little bit, I need to do more perf and bench to verify. 🤣

Reply:
I tested this locally, also getting better performance after the change(!).

It suggests we are doing something wrong maybe with locking (e.g. stalling some work), or we somehow perform less work in this PR? I didn't check whether we still get the right results yet.

Reply:
Turns out it was some noise - can not reproduce this.

Reply:
@Dandandan That's weird. I've tested it several times, and it hasn't changed 🤦‍♂️ -- I'll close the ticket and work on it when I have time.